### ISSUE TITLE
CSS custom properties

### ISSUE THREAD

**Issue by o-t-w** [frenic/csstype#63]
I am writing an app in React and Typescript and will be making heavy use of CSS custom properties.

Could you idiot-proof the documentation a bit more please? Is `css.d.ts` a file I need to create and does it matter what directory it is in?  
```
// My css.d.ts file
import * as CSS from 'csstype';

declare module 'csstype' {
  interface Properties {
    // Add a missing property
    WebkitRocketLauncher?: string;

    // Add a CSS Custom Property
    '--theme-color'?: 'black' | 'white';

    // ...or allow any other property
    [index: string]: any;
  }
}
```

**Comment by rlacorne** [comment]
I would like to know as well.

I don't think augmenting the module should be the way to go, as the standard '--string: any' CSS custom props syntax should be supported by default, but as of now, augmenting the module seems to be the only way, and I cannot get it to work.

**Comment by o-t-w** [comment]
CSS custom properties are a major web feature that will only increase in popularity as Internet Explorer dies. Needing such a large amount of boilerplate in _every single module_ that uses them is enough to make me abandon Typescript altogether. They really do need to work out of the box.

**Comment by frenic** [comment]
It actually up to vendors to decide if they want to support CSS Custom Properties by default or not. But the only way is to support them out of the box is to have index signatures (`[index: string]: any;`) and the a big reason multiple libraries migrated to CSSType was to be able to get rid of them. You can't have index signatures for a specific pattern like properties that starts with `--` yet I'm afraid and I'm unsure if it ever will be supported.

```ts
// My css.d.ts file
import * as CSS from 'csstype';

declare module 'csstype' {
  interface Properties {
    [index: string]: any;
  }
}
```

In tsconfig.json you make sure it's included:

```json
{
  "include": [
    "path/to/css.d.ts"
  ]
}
```

But some libraries doesn't extend `CSS.Properties` and in those cases this won't work. You could try to augment `CSS.PropertiesFallback` instead which is fairly common too.

```ts
// My css.d.ts file
import * as CSS from 'csstype';

declare module 'csstype' {
  interface PropertiesFallback {
    [index: string]: any;
  }
}
```

Otherwise you'l have to augment the style interface in that specific library. Like this example with Glitz:

```ts
// My glitz.d.ts file
import * as Glitz from '@glitz/type';

declare module '@glitz/type' {
  interface Style {
    [index: string]: any;
  }
}
```

Either way, I would say the it's up to each library to have their own recommended way of supporting CSS Custom Properties. So raise a ticket to whatever library you're using for styling purposes if this answer doesn't satisfy you.

**Comment by jacobrask** [comment]
It would be unfortunate if any `--` property would be valid, because a major use case is to explicitly only allow the custom properties you have defined in a theme for instance to avoid typos.

**Comment by yordis** [comment]
I will recommend you to do the following strategy.

```ts
import * as React from 'react';

export interface AppStyle extends React.CSSProperties {
  '--app-background-color': string;
  '--app-color': string;
}

export const App: React.FunctionComponent<IAppProp> = props => {
  const style: AppStyle = {
    '--app-background-color': props.backgroundColor,
    '--app-color': props.color,
  };

  return (
    <div style={style}></div>
  );
};
```

**Comment by o-t-w** [comment]
@frenic 
> You can't have index signatures for a specific pattern like properties that starts with `--` yet I'm afraid and I'm unsure if it ever will be supported.

Where can I file an issue to get this sorted out?

@jacobrask 
> It would be unfortunate if any `--` property would be valid, because a major use case is to explicitly only allow the custom properties you have defined in a theme for instance to avoid typos.

Not sure I understand this comment. Most people just define their custom properties in a stylesheet.

**Comment by jacobrask** [comment]
You would write an interface with the custom variables in use in your code base. It would have to be duplicated across CSS and TypeScript. Just like you define interfaces for APIs etc.

**Comment by frenic** [comment]
Hopefully https://github.com/microsoft/TypeScript/issues/6579 will become reality soon and we could do something like:

```ts
interface Properties {
    // ....
    [customProperty: /--\w+/]: number | string;
}
```

**Comment by cyrilchapon** [comment]
But still; even with a pattern-aware Typescript, this won't be a good thing IMHO to allow "any of -- starting index" at the root of every user style.

Custom properties are a major feature, agreed; it doesn't mean it has to be opted in by default and not-typed and monkey accepted in any CSS in JS stylesheet. They key point is @jacobrask one. At the end of the day a custom property is a - dynamic - property; just like any property in any Typescript type. Csstype provides **standard** properties; one need a custom one, he should add it to its own interface (extending csstype ones)

**Comment by cyrilchapon** [comment]
I'll rephrase by example :

**TODAY**
```
import { Properties } from 'csstype'

interface Style extends CSS.Properties {}

const style: Style = {
  color: 'red', // <= ✅OK
  CoLoR: 'purple', // <= ✅Type Error
  -myColor: 'orange', // <= ⛔️Type Error (what you want to accept)
  -MyCoLoR: 'green' // <= ✅Type Error
}
```

**With suggested interface change**
```
import { Properties } from 'csstype'

interface Style extends CSS.Properties {
  [index: string]: any
}

const style: Style = {
  color: 'red', // <= ✅OK
  CoLoR: 'purple', // <= ⛔️OK
  -myColor: 'orange', // <= ✅OK
  -MyCoLoR: 'green' // <= ⛔️OK
}
```

**With ideal pattern-aware-Typescript interface change**
** ⚠️this is pseudo code; Typescript doesn't have this feature for now **
```
import { Properties } from 'csstype'

interface Style extends CSS.Properties {
  // Imagine
  [index: StartsWith<string, '--'>]: any
}

const style: Style = {
  color: 'red', // <= ✅OK
  CoLoR: 'purple', // <= ✅Type error
  -myColor: 'orange', // <= ✅OK
  -MyCoLoR: 'green' // <= ⛔️OK
}
```

The only way to get correct type-checking across your app is to manually add any custom property you want to accept.

**Solution**
```
import { Properties } from 'csstype'

interface Style extends CSS.Properties {
  --myColor?: string
}

const style: Style = {
  color: 'red', // <= ✅OK
  CoLoR: 'purple', // <= ✅Type Error
  -myColor: 'orange', // <= ✅OK (your property, accepted)
  -MyCoLoR: 'green' // <= ✅Type Error
}
```


As for anything in typescript; if your "custom property set" is generic; well go with Typescript generics.

**Solution with generic**
```
import { Properties } from 'csstype'

interface StyleWithCustomProperties<T> extends CSS.Properties {
  [customPropKey: keyof T]?: T[customPropKey]
}


interface CustomProperties {
  --myColor?: string
}

const style: Style<CustomProperties> = {
  color: 'red', // <= ✅OK
  CoLoR: 'purple', // <= ✅Type Error
  -myColor: 'orange', // <= ✅OK (your property, accepted)
  -MyCoLoR: 'green' // <= ✅Type Error
}
```
(this last one is untested pseudo code, but should work)

**Comment by frenic** [comment]
I agree with @cyrilchapon. Custom properties are what they are, custom properties. And it's impossible to automatically know **your** custom properties. So end users are forced to specify them somehow.

However, CSSType is mainly an utility for third party libraries to provide types for known CSS properties. Third party libraries are the ones who decides how/if custom properties are supported and therefor also the ones who decides how they are typed. So I don't consider CSSType the be responsible to solve this issue. Even if we add a generic to ease things up a bit. It will likely not benefit the end user.

I understand that it's painful when third libraries that supports Custom Properties as is and are poorly typed. In that case, the best way to manually specify custom properties using Typescript is by augmentation as [described earlier](https://github.com/frenic/csstype/issues/63#issuecomment-451939423). The issue today is that the end users has to worry about which interface of CSSType their third party style library uses. But we can get around that if all style interfaces extends an empty `CustomProperties` interface that can be augmented. Example:

```ts
// My css.d.ts file
import * as CSS from 'csstype';

declare module 'csstype' {
  interface CustomProperties {
    --myColor: string;
    --myBackground: string;
  }
}
```

```json
// My tsconfig.json file
{
  "include": [
    "path/to/css.d.ts"
  ]
}
```
```ts
// My style.ts file
import * as CSS from 'csstype';

const style: CSS.Properties = {
  '--myColor': '#FF0000', <-- OK
  '--myBackground': 1, <-- Fail, should be a string
  '--myPadding': 10, <-- Fail, unknown property
}

const fallbackStyle: CSS.FallbackProperties = {
  '--myColor': ['#FF0000AA', '#FF0000'], <-- OK
  '--myBackground': 1, <-- Fail, should be a string
  '--myPadding': 10, <-- Fail, unknown property
}
```

**Comment by cobalt-medma** [comment]
What if I wanted to modify the type definition of a prop, for example "fontWeight"? 
I tried the following but it's not working. Maybe because of my lack of typescript knowledge.

```ts
import * as CSS from 'csstype'

declare module 'csstype' {
  export interface Properties {
    fontWeight: Property.FontWeight | 'medium'
  }
}

```

**Comment by mohsen1** [comment]
TypeScript 4.4 does allow specifying types for properties with a pattern. In this case you can specify something like this:

```ts
interface CSSProperties {
  // ... all other properties
  [variableName: `--${string}`]: any;
}
```

**Comment by nahumzs** [comment]
What I ended doing.

// types.ts
```ts
import { Properties } from "csstype";

export interface CSSProperties extends Properties {
  [variableName: `--${string}`]: any;
}
```

```tsx
import { CSSProperties } from "./types";

  const style: CSSProperties = {
    "--width": "10px",
    "--height": "10px",
  };

  return <Component style={style} />;

  ```
